### PR TITLE
EVMHost: charge gas for precompiles

### LINKED ISSUE #11057: [isoltest] Charge proper gas for precompiles in EVMHost
Once we have #10474 we'll have a lot of opportunity to compare gas usage of optimised vs non-optimised and legacy vs. ir-based pipelines. With #10475 we'll have the opportunity to insert more realistic external tests, like we did with #10590.

At that point, having more realistic gas costs for precompiles could be useful. Especially for ecrecover and sha256 as they are used in external tests and have simple clear costs. Other precompiles are more complicated and not well used in our test suite so of less concern.

Currently we [charge 0 gas](https://github.com/ethereum/solidity/blob/develop/test/EVMHost.cpp#L352) for them.

### PR DESCRIPTION
**Pull request by axic**
Closes #11057.

### CODE REVIEW

**Comment by leonardoalt** [comment]
This PR looks small so we could get it done. What's missing?
@axic

**Comment by leonardoalt** [comment]
Marked as `takeover` and no activity, closing the PR. Code is still available in the branch if anyone wants to take over.

**Comment by cameel** [comment]
We decided to deprecate the gas meter anyway.

**Comment by axic** [comment]
> We decided to deprecate the gas meter anyway.

I think this is not related. Wasn't the decision to remove the gas estimator? This here practically is part of evmone and testing, and is useful to get accurate gas figures in the tests.

**Comment by cameel** [comment]
My bad. I must have misinterpreted it back then. I thought it was about handling precompiles in the meter but obviously not since this is about EVMHost.

**Comment by ekpyron** [comment]
So what's the state here? I guess I'll add back the ``takeover`` label?

**Comment by axic** [comment]
> So what's the state here? I guess I'll add back the takeover label?

No need, still trying to figure out, there's some memory corruption happening which is hard to debug.

**Comment by axic** [comment]
```
Expected result:
// f(int256): 4 -> 0x1b0f3c404d12075c68c938f9f60ebea4f74941a0000000000000000000000000
// f(int256): 5 -> 0xee54aa84fc32d8fed5a5fe160442ae84626829d9000000000000000000000000
// f(int256): -1 -> 0x1cf4e77f5966e13e109703cd8a0df7ceda7f3dc3000000000000000000000000

Obtained result:
// f(int256): 4 -> 0xd35500004d12075c68c938f9f60ebea4f74941a0000000000000000000000000
Warning: The call to "f(int256)" returned 
[d3,55,0,0,4d,12,7,5c,68,c9,38,f9,f6,e,be,a4,f7,49,41,a0,0,0,0,0,0,0,0,0,0,0,0,0]
// f(int256): 5 -> 0xd3550000fc32d8fed5a5fe160442ae84626829d9000000000000000000000000
Warning: The call to "f(int256)" returned 
[d3,55,0,0,fc,32,d8,fe,d5,a5,fe,16,4,42,ae,84,62,68,29,d9,0,0,0,0,0,0,0,0,0,0,0,0]
// f(int256): -1 -> 0xd35500005966e13e109703cd8a0df7ceda7f3dc3000000000000000000000000
Warning: The call to "f(int256)" returned 
[d3,55,0,0,59,66,e1,3e,10,97,3,cd,8a,d,f7,ce,da,7f,3d,c3,0,0,0,0,0,0,0,0,0,0,0,0]
```

For example here, that first 32-bits are overwritten by something and I can't figure it out. The fields are not overlapping. I was surmising maybe the `map<.., tuple<>>` construct is the problem.

File: test/EVMHost.cpp [hunk 1]
```diff
@@ -930,19 +1105,44 @@ evmc::result EVMHost::precompileALTBN128PairingProduct(evmc_message const& _mess
...
 evmc::result EVMHost::precompileGeneric(
 	evmc_message const& _message,
-	map<bytes, bytes> const& _inOut) noexcept
+	map<bytes, tuple<bytes, int64_t>> const& _inOut) noexcept
 {
 	bytes input(_message.input_data, _message.input_data + _message.input_size);
 	if (_inOut.count(input))
-		return resultWithGas(_message, _inOut.at(input));
+	{
+		auto ret = _inOut.at(input);
```
Comment (axic) [inline test/EVMHost.cpp]: This is making a copy, this is messing it up. Plus somehow not `T const& get<>` is used.

**Comment by axic** [comment]
@cameel @nikola-matic this should be easy enough to review/merge however

File: test/EVMHost.cpp [hunk 2]
```diff
@@ -418,15 +440,19 @@ evmc::result EVMHost::precompileECRecover(evmc_message const& _message) noexcept
...
+				fromHex("0000000000000000000000008743523d96a1b2cbe0c6909653a56da18ed484af"),
+				gas_cost
+			}
 		}
 	};
 	evmc::result result = precompileGeneric(_message, inputOutput);
 	// ECRecover will return success with empty response in case of failure
-	if (result.status_code != EVMC_SUCCESS)
+	if (result.status_code != EVMC_SUCCESS && result.status_code != EVMC_OUT_OF_GAS)
+//		return resultWithGas(_message.gas, gas_cost, {});
```
Comment (cameel) [inline test/EVMHost.cpp:452]: Doesn't look like this belongs here given that we want result to be success in this case.

Please remove or add a comment explaining why it's here.

File: test/EVMHost.cpp [hunk 3]
```diff
@@ -418,15 +440,19 @@ evmc::result EVMHost::precompileECRecover(evmc_message const& _message) noexcept
...
+			{
+				fromHex("0000000000000000000000008743523d96a1b2cbe0c6909653a56da18ed484af"),
+				gas_cost
+			}
 		}
 	};
 	evmc::result result = precompileGeneric(_message, inputOutput);
 	// ECRecover will return success with empty response in case of failure
-	if (result.status_code != EVMC_SUCCESS)
+	if (result.status_code != EVMC_SUCCESS && result.status_code != EVMC_OUT_OF_GAS)
```
Comment (cameel) [inline test/EVMHost.cpp:451]: Looks like `EVMC_OUT_OF_GAS` case is a bug fix? I.e. before it would be replaced by success, which was incorrect?

I'd put it in a separate commit if that's the case here.

**Review by cameel — APPROVED** [review]
I have two cosmetic suggestions but overall the behavior and costs in the PR seem correct.

File: test/EVMHost.cpp (hunk 3 above)
Comment (axic) [reply test/EVMHost.cpp:451]: Not a bugfix, just that previously it made no difference as gas was not charged. Now it needs to be handled in a separate case, because of the odd behaviour of ecrecover: consumes the same fixed gas in both success and error cases (unless there was not enough gas available to begin with).

File: test/EVMHost.cpp (hunk 2 above)
Comment (axic) [reply test/EVMHost.cpp:452]: I actually wanted to use this but then had concerns the new `{}` object used as `const&` could cause issues. If it is fine we can use that shorthand, otherwise just drop the line.

File: test/EVMHost.cpp (hunk 3 above)
Comment (cameel) [reply test/EVMHost.cpp:451]: Ah, you mean that previously out of gas would actually never happen here? Makes sense.

File: test/EVMHost.cpp (hunk 2 above)
Comment (cameel) [reply test/EVMHost.cpp:452]: Ah, right. Just realized that the function sets success as well so it's equivalent after all (I've been staring at this code too long :)).

Yeah, the `const&` might be a concern - even if not an actual problem, we had cases of sanitizers complaining about that.

In any case, it's fine for me either way, just with a short comment that it's equivalent if you leave this commented out.

Comment (axic) [reply test/EVMHost.cpp:452]: I'd just remove it, but also don't want to wait for CI. If we are confident that #13532 is merged soon we could remove the commented out line there? 😅

Comment (ekpyron) [reply test/EVMHost.cpp:452]: Not sure I understand the concern here - ``{}`` bound to a ``const&`` is a temporary that lives for the lifetime of the statement, so as long as it's copied on the other end of the call, this is perfectly fine. I actually don't see a matching ``resultWithGas`` overload, though :-). Anyways, fine with adjusting this in #13532, which we *will* merge soon, so merging ;-).
